library/src/main/java/com/datatorrent/lib/state/managed/Bucket.java

```diff
@@ -0,0 +1,534 @@
+    private final transient TreeMap<Long, Map<Slice, BucketedValue>> checkpointedData = Maps.newTreeMap();
+
+    //Data persisted in bucket data files
+    private final transient Map<Slice, BucketedValue> committedData = Maps.newHashMap();
+
+    //Data recovered
+    private final transient TreeMap<Long, Map<Slice, BucketedValue>> recoveredData = Maps.newTreeMap();
+
+    //Data serialized/deserialized from bucket data files
+    private final transient Map<Slice, BucketedValue> readCache = Maps.newConcurrentMap();
+
+    //TimeBucket -> FileReaders
+    private final transient Map<Long, FileAccess.FileReader> readers = Maps.newTreeMap();
+
+    protected transient ManagedStateContext managedStateContext;
+
+    private AtomicLong sizeInBytes = new AtomicLong(0);
+
+    private final transient Slice dummyGetKey = new Slice(null, 0, 0);
+
+    private transient TreeSet<BucketsFileSystem.ImmutableTimeBucketMeta> cachedBucketMetas;
+
+    private DefaultBucket()
+    {
+      //for kryo
+      bucketId = -1;
+    }
+
+    protected DefaultBucket(long bucketId)
+    {
+      this.bucketId = bucketId;
+    }
+
+    @Override
+    public void setup(@NotNull ManagedStateContext managedStateContext)
+    {
+      this.managedStateContext = Preconditions.checkNotNull(managedStateContext, "managed state context");
+    }
+
+    @Override
+    public long getBucketId()
+    {
+      return bucketId;
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+      return sizeInBytes.longValue();
+    }
+
+    private Slice getFromMemory(Slice key)
+    {
+      //search the cache for key
+      BucketedValue bucketedValue = flash.get(key);
+      if (bucketedValue != null) {
+        return bucketedValue.getValue();
+      }
+
+      for (Long window : checkpointedData.descendingKeySet()) {
+        //traverse the checkpointed data in reverse order
+        bucketedValue = checkpointedData.get(window).get(key);
+        if (bucketedValue != null) {
+          return bucketedValue.getValue();
+        }
+      }
+
+      for (Long window : recoveredData.descendingKeySet()) {
+        //traverse the reccovered data in reverse order
+        bucketedValue = recoveredData.get(window).get(key);
+        if (bucketedValue != null) {
+          return bucketedValue.getValue();
+        }
+      }
+
+      bucketedValue = committedData.get(key);
+      if (bucketedValue != null) {
+        return bucketedValue.getValue();
+      }
+
+      bucketedValue = readCache.get(key);
+      if (bucketedValue != null) {
+        return bucketedValue.getValue();
+      }
+      return null;
+    }
+
+    private Slice getFromReaders(Slice key, long timeBucket)
+    {
+      if (timeBucket != -1) {
+        Slice valSlice = getKeyFromTimeBucketReader(key, timeBucket);
+        if (valSlice != null) {
+          BucketedValue bucketedValue = new BucketedValue(timeBucket, valSlice);
+          readCache.put(key, bucketedValue);
```

Comment:
Will fix this. However the expectation here is that the latest value is retrieved always. If the latest time-bucket is known then the process to read it is faster otherwise reading all time-buckets is slower. This is again missing from the documentation  and I will add to it.